### DISCUSSION TITLE
How can I get the minor currency unit with rounding?

### DISCUSSION THREAD

**Discussion by DustinJSilk** [dinerojs/dinero.js#513] [Q&A] [answered]
After doing a few calculations, I've found my scale is larger than the currencies original scale.
This means that when I create a snapshot to get the amount, the amount doesn't actually represent the currencies minor value. And the toUnit method can only safely return a major unit.

For example:

```TypeScript
const cost = dinero({ amount: 1000, currency: USD });
const taxed = multiply(cost, { amount: 11256, scale: 4 });

const {
  amount, // 11256000
  scale, // 6
  currency,
} = toSnapshot(taxed);
```

In this case, I'd need to send stripe 1126, not 112560. 
If I use `transformScale`, I get 1125, not 1126.

This is the only way I could figure it out:

```TypeScript
const unit = toUnit(taxed, { digits: currency.exponent, round: up });
const minorFromUnit = unit * Math.pow(currency.base, currency.exponent)
console.log(minorFromUnit) // 1126
```

A `toMinorUnit` would be extremely helpful in this type of situation.
Or potentially giving the `toUnit` function a scale option.
Or even give a rounding option to transformScale. This way we can easily save snapshots in a database without worrying about saving an amount that could be different to the amount charged to customers.

Appreciate the effort! 
Thanks

**Comment by tranhl** [comment]
@DustinJSilk FYI the solution you provided is potentially erroneous as the float multiplication without rounding can create imprecise values (e.g. `6581.99999999`), so you'll have to apply a rounding mode after the multiplication:

```ts
const minorFromUnit = halfUp(unit * Math.pow(currency.base, currency.exponent))
```

**Comment by sarahdayan** [comment]
Hey @DustinJSilk and sorry for the late reply.

You're right, there currently isn't a good way to transform an object with loss and specify how the loss is handled—[`transformScale` just truncates the amount](https://v2.dinerojs.com/docs/api/conversions/transform-scale). I've thought about it for a while, and I'm leaning towards providing a userland rounding function to `transformScale`.

If you'd like to try it out, a PR is always welcome!

>This way we can easily save snapshots in a database without worrying about saving an amount that could be different to the amount charged to customers.

I would generally recommend persisting accurate, precise snapshots, and saving rounded amounts separately. Whenever you need to do further calculations, it's usually nice to have the raw, unatered data saved somewhere.

**Comment by sarahdayan** [comment, marked as answer]
Update: `transformScale` now accepts an optional divide function as its third argument, which controls how rounding is handled when scaling down. It defaults to `down` (truncation), but you can pass `halfUp`, `up`, or any other rounding mode:

```js
import { dinero, multiply, transformScale, toSnapshot, up } from 'dinero.js';
import { USD } from 'dinero.js/currencies';

const cost = dinero({ amount: 1000, currency: USD });
const taxed = multiply(cost, { amount: 11256, scale: 4 });

const rounded = transformScale(taxed, 2, up);
toSnapshot(rounded).amount; // 1126
```

This is exactly the feature discussed in this thread.
